src/meetingreadinesschecker/DefaultMeetingReadinessChecker.ts

```diff
@@ -0,0 +1,451 @@
+      let userFeedback = await callback();
+      this.stopTone();
+      if (userFeedback) {
+        return CheckAudioOutputFeedback.Success;
+      } else {
+        return CheckAudioOutputFeedback.Failure;
+      }
+    } catch (e) {
+      this.logger.error(`checkAudioOutput failed with error: ${e}`);
+      return CheckAudioOutputFeedback.Failure;
+    }
+  }
+
+  playTone(sinkId: string | null, frequency: number | 440): void {
+    const rampSec = 0.1;
+    const maxGainValue = 0.1;
+
+    let destinationStream: MediaStreamAudioDestinationNode;
+    this.audioContext = DefaultDeviceController.getAudioContext();
+    this.gainNode = this.audioContext.createGain();
+    this.gainNode.gain.value = 0;
+    this.oscillatorNode = this.audioContext.createOscillator();
+    this.oscillatorNode.frequency.value = frequency;
+    this.oscillatorNode.connect(this.gainNode);
+    destinationStream = this.audioContext.createMediaStreamDestination();
+    this.gainNode.connect(destinationStream);
+    const currentTime = this.audioContext.currentTime;
+    const startTime = currentTime + 0.1;
+    this.gainNode.gain.linearRampToValueAtTime(0, startTime);
+    this.gainNode.gain.linearRampToValueAtTime(maxGainValue, startTime + rampSec);
+    this.oscillatorNode.start();
+    const audioMixController = new DefaultAudioMixController();
+    // @ts-ignore
+    audioMixController.bindAudioDevice({ deviceId: sinkId });
+    audioMixController.bindAudioElement(new Audio());
+    audioMixController.bindAudioStream(destinationStream.stream);
+  }
+
+  stopTone(): void {
+    const durationSec = 1;
+    const rampSec = 0.1;
+    const maxGainValue = 0.1;
+    const currentTime = this.audioContext.currentTime;
+    this.gainNode.gain.linearRampToValueAtTime(maxGainValue, currentTime + rampSec + durationSec);
+    this.gainNode.gain.linearRampToValueAtTime(0, currentTime + rampSec * 2 + durationSec);
+    this.oscillatorNode.stop();
+  }
+
+  async checkVideoInput(videoInputDeviceInfo: MediaDeviceInfo): Promise<CheckVideoInputFeedback> {
+    try {
+      const result = await this.meetingSession.audioVideo.chooseVideoInputDevice(
+        videoInputDeviceInfo
+      );
+      if (
+        result === DevicePermission.PermissionDeniedByBrowser ||
+        result === DevicePermission.PermissionDeniedByUser
+      ) {
+        return CheckVideoInputFeedback.PermissionDenied;
+      } else {
+        await this.meetingSession.audioVideo.chooseVideoInputDevice(null);
+        return CheckVideoInputFeedback.Success;
+      }
+    } catch (error) {
+      this.logger.error(`MeetingReadinessChecker: Video check failed with error ${error}`);
+      return CheckVideoInputFeedback.Failure;
+    }
+  }
+
+  async checkCameraResolution(
+    videoInputDevice: MediaDeviceInfo,
+    width: number,
+    height: number
+  ): Promise<CheckCameraResolutionFeedback> {
+    const videoConstraint = {
+      video: { width: { exact: width }, height: { exact: height } },
```

Comment:
- Using the `exact` keyword will always fail on the early versions of iOS Safari 12. See this commit (https://github.com/aws/amazon-chime-sdk-js/commit/fe5235778713b7ea2ebaf1a078b5ed428ecf7a6b#diff-44c2a21034320b3414d79247fbd51454R738) that we are conditionally using the `exact` keyword for the `deviceId` field. I'm not sure with `width` and `height` fields.

- Also, see this PR that we needed to use the `ideal` keyword for `width` and `height`. PR says `some mobile browsers`. 
  https://github.com/aws/amazon-chime-sdk-js/pull/540

Note that if we loose constraints, those browsers support camera resolutions.

Reply:
I tried ideal but it seems to always succeed since it just looks for the closest resolution.